libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,831 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  implementation of RunCam camera protocols 
+
+  With thanks to betaflight for a great reference
+  implementation. Several of the functions below are based on
+  betaflight equivalent functions
+
+  RunCam protocol specification can be found at https://support.runcam.com/hc/en-us/articles/360014537794-RunCam-Device-Protocol
+ */
+#include "AP_RunCam.h"
+
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Logger/AP_Logger.h>
+#include <AP_OSD/AP_OSD.h>
+
+const AP_Param::GroupInfo AP_RunCam::var_info[] = {
+    // @Param: FEATURES
+    // @DisplayName: RunCam features available
+    // @Description: The available features of the attached RunCam device. This should be determined automatically but can be set manually.
+    // @User: Advanced
+    AP_GROUPINFO("FEATURES", 0, AP_RunCam, _features, 0),
+
+    // @Param: PROT_VER
+    // @DisplayName: RunCam protocol Version
+    // @Description: The protocol version supported by the attached RunCam device. Only protocol version 1 is supported.
+    // @User: Advanced
+    AP_GROUPINFO("PROT_VER", 1, AP_RunCam, _protocol_version, RCDEVICE_PROTOCOL_VERSION_1_0),
+
+    // @Param: INIT_NUM
+    // @DisplayName: RunCam init attempts
+    // @Description: Number of times that the attached RunCam device will be attempted to be initialised.
+    // @User: Advanced
+    AP_GROUPINFO("INIT_NUM", 2, AP_RunCam, _init_attempts, 6),
+
+    // @Param: INIT_INT
+    // @DisplayName: RunCam init interval
+    // @Description: Interval between RunCam initialisation attempts in ms.
+    // @User: Advanced
+    AP_GROUPINFO("INIT_INT", 3, AP_RunCam, _init_attempt_interval_ms, 1000),
+
+    // @Param: BT_DELAY
+    // @DisplayName: RunCam boot delay before allowing updates
+    // @Description: Time it takes for the RunCam to become fully ready in ms. If this is too short then commands can get out of sync.
+    // @User: Advanced
+    AP_GROUPINFO("BT_DELAY", 4, AP_RunCam, _boot_delay_ms, 6000),
+
+    // @Param: BTN_DELAY
```

Comment:
Ok we definitely need this. There is no way to tell when the camera has booted and is accepting requests other than waiting. Wait too short a time and everything gets out of sync, too long and its annoying. I am sure this will vary depending on camera version and power setup.